src/metabase/db/query.clj

```diff
@@ -22,19 +22,80 @@
   (:require
    [clojure.string :as str]
    [honey.sql :as sql]
-   [metabase.db.connection :as mdb.connection]
+   [metabase.db :as mdb]
    [metabase.driver :as driver]
    [metabase.plugins.classloader :as classloader]
+   [metabase.util :as u]
    [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.schema :as ms]
    [toucan2.core :as t2]
-   [toucan2.jdbc.options :as t2.jdbc.options]))
+   [toucan2.jdbc.options :as t2.jdbc.options]
+   [toucan2.model :as t2.model])
+  (:import
+   (clojure.lang ExceptionInfo)))
 
 (set! *warn-on-reflection* true)
 
 (defn format-sql
   "Return a nicely-formatted version of a `query` string with the current application db driver formatting."
   [sql]
-  (driver/prettify-native-form (mdb.connection/db-type) sql))
+  (driver/prettify-native-form (mdb/db-type) sql))
+
+(def ^:private NamespacedKeyword
+  [:and :keyword [:fn (comp seq namespace)]])
```

Comment:
that's true but there are implications about whether it needs a nice error message. I'm going to punt on that and let another pass clean it up.

Reply:
You can just use `clojure.core/qualified-keyword?` directly as a schema and it will do the right thing.

```clj
(malli.error/humanize (malli.core/explain qualified-keyword? :abc))
;; => ["should be a qualified keyword"]
```

If you write a `:fn` schema, you should include an :error/message` in the options, otherwise if that check fails it's going to give you a meaningless explanation:

```clj
(malli.error/humanize (malli.core/explain NamespacedKeyword :abc))
;; => ["unknown error"]

(def ^:private NamespacedKeyword2
  [:and :keyword [:fn {:error/message "should be a namespaced keyword"} (comp seq namespace)]])

(malli.error/humanize (malli.core/explain NamespacedKeyword2 :abc))
;; => ["should be a namespaced keyword"]
```